pkg/service/keystore/storage.go

```diff
@@ -48,43 +56,107 @@ const (
 )
 
 type Storage struct {
-	db         storage.ServiceStorage
-	serviceKey []byte
+	db        storage.ServiceStorage
+	encryptor Encryptor
+	decryptor Decryptor
 }
 
-func NewKeyStoreStorage(db storage.ServiceStorage, key ServiceKey) (*Storage, error) {
-	keyBytes, err := base58.Decode(key.Base58Key)
+func NewKeyStoreStorage(db storage.ServiceStorage, e Encryptor, d Decryptor) (*Storage, error) {
+	bolt := &Storage{
+		db:        db,
+		encryptor: e,
+		decryptor: d,
+	}
+
+	return bolt, nil
+}
+
+type wrappedEncryptor struct {
+	tink.AEAD
+}
+
+func (w wrappedEncryptor) Encrypt(_ context.Context, plaintext, contextData []byte) ([]byte, error) {
+	return w.AEAD.Encrypt(plaintext, contextData)
+}
+
+type wrappedDecryptor struct {
+	tink.AEAD
+}
+
+func (w wrappedDecryptor) Decrypt(_ context.Context, ciphertext, contextInfo []byte) ([]byte, error) {
+	return w.AEAD.Decrypt(ciphertext, contextInfo)
+}
+
+const (
+	gcpKMSScheme = "gcp-kms"
+	awsKMSScheme = "aws-kms"
+)
+
+func NewEncryption(db storage.ServiceStorage, cfg config.KeyStoreServiceConfig) (Encryptor, Decryptor, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if strings.HasPrefix(cfg.MasterKeyURI, gcpKMSScheme) || strings.HasPrefix(cfg.MasterKeyURI, awsKMSScheme) {
+		return NewExternalEncryptor(ctx, cfg)
+	}
+
+	// First, generate a service key
+	serviceKey, serviceKeySalt, err := GenerateServiceKey(cfg.ServiceKeyPassword)
 	if err != nil {
-		return nil, errors.Wrap(err, "could not decode service key")
+		return nil, nil, sdkutil.LoggingErrorMsg(err, "generating service key")
 	}
-	bolt := &Storage{db: db, serviceKey: keyBytes}
 
-	// first, store the service key
-	if err = bolt.storeServiceKey(context.Background(), key); err != nil {
-		return nil, errors.Wrap(err, "could not store service key")
+	key := ServiceKey{
+		Base58Key:  serviceKey,
+		Base58Salt: serviceKeySalt,
 	}
-	return bolt, nil
+	if err := storeServiceKey(ctx, db, key); err != nil {
+		return nil, nil, err
+	}
+	return &encryptor{db}, &decryptor{db}, nil
+}
+
+func NewExternalEncryptor(ctx context.Context, cfg config.KeyStoreServiceConfig) (Encryptor, Decryptor, error) {
```

Comment:
nit: here and elsewhere `encryptor` and `decryptor` are british spellings, prefer `encrypter` and `decrypter`